terraform/modules/slack_alert_lambda/code/slack_alerts.rb

```diff
@@ -0,0 +1,324 @@
+require 'json'
+require 'net/https'
+require 'logger'
+
+def send_alert(event)
+  logger = Logger.new($stdout)
+
+  logger.info("Event recieved: #{event}")
+
+  message = construct_message(event)
+  response = send_message(message)
+
+  logger.info("Sent slack message. Response: #{response.code} #{response.body}")
+
+  # Should raise if not success
+  response.value
+end
+
+def construct_message(event)
+  return alarm_message(event) if is_alarm?(event)
+  return budget_alert_message(event) if is_budget_alert?(event)
+  return ecr_scan_result_message(event) if is_ecr_scan_result?(event)
+
+  unknown_alert_message(event)
+end
+
+def is_alarm?(event)
+  event_message(event).match?('AlarmName')
+end
+
+def alarm_message(event)
+  alarm = parse_alarm(event)
+  message_for_alarm(alarm)
+end
+
+def parse_alarm(event)
+  message = JSON.parse(event_message(event))
+  {
+    name: message['AlarmName'],
+    description: message['AlarmDescription'],
+    reason: message['NewStateReason'],
+    state: message['NewStateValue'],
+    previous_state: message['OldStateValue'],
+    time: message['StateChangeTime']
+  }
+end
+
+def message_for_alarm(alarm)
+  case alarm[:state]
+  when 'ALARM'
+    alarm_activated_message(alarm)
+  when 'OK'
+    alarm_resolved_message(alarm)
+  when 'INSUFFICIENT_DATA'
+    # We set notifications not to be sent to sns when we don't care about these
+    alarm_activated_message(alarm)
+  else
+    raise "Unrecognised alarm state #{alarm[:state]}"
+  end
+end
+
+def alarm_activated_message(alarm)
+  {
+    blocks: [
+      { 
+        type: 'header',
+        text: {
+          type: 'plain_text',
+          text: "#{alarming_slack_emoji} #{ENV['ENVIRONMENT']} alarm triggered: #{alarm[:name]}"
```

Comment:
Is AlarmName something like `core-prod-cloudfront-5xx-error-rate` ? I'm on the fence on whether it's pleasant enough to see in a header. Would it be a pain to map it to something more human?

Reply:
It is - and I think we need it in there somewhere to help find things on AWS, although it doesn't have to be the header. We could probably map it, or maybe use some string manipulation to get e.g. "Cloudfront 5xx Error Rate"?

Might depend how much forms work we have ready!